Add editor in reader behind a feature flag

creativecoder opened the pull request:
## Proposed Changes

Adds an instance of isolated-block-editor behind a feature flag, `reader/editor`.

This is meant to be a starting point that we can build on. See pe7F0s-Kk-p2

## Testing Instructions

* Open `/read` on calypso.localhost
* Click on the editor and create a post.
* Click publish to publish the post on your primary site.

tyxla commented on package.json:
```diff
@@ -304,7 +304,10 @@
-		"@wordpress/element": "4.20.0"
+		"@wordpress/element": "4.20.0",
+		"@wordpress/block-editor": "10.4.0",
+		"@wordpress/block-library": "7.17.0",
```
I wonder if we could pin to the versions we use broadly right now:

* `@wordpress/bock-editor@10.5.0`
* `@wordpress/block-library@7.19.0`

The other forced resolutions here seem to match what we're already using at this time.

This is, of course, temporary, since as @noahtallen mentioned in p4TIVU-auj-p2 we are in the middle of upgrading Calypso to the latest `@wordpress` packages and upgrading to React 18 in the meantime.

creativecoder replied:
Unfortunately, changing to those versions starts to trigger console errors--there are probably other package version changes that would need to be in a new version of isolated-block-editor to align with those versions of block-editor and block-library.

Using `wordpress/block-editor@10.5.0` and `@wordpress/block-library@7.19.0`, here's the error I see, when using slash to insert a block: 2ff1d-pb. I also tried changing just `wordpress/block-editor@10.5.0` (with `@wordpress/block-library@ 7.17.0`), but that results in a build error.

tyxla replied:
Makes sense. I hope we'll get to address most of these issues as we upgrade (and maintain) Calypso and all related projects to the latest version of the `@wordpress` packages eventually.

tyxla commented on config/production.json:
```diff
@@ -119,6 +119,7 @@
 		"purchases/new-payment-methods": true,
 		"push-notifications": true,
 		"reader": true,
+		"reader/editor": false,
```
Note that a feature flag is always `false` by default, unless specifically set to `true`. That being said, a `false` flag can just be omitted and it will work the same way.

creativecoder replied:
Good point--usually I err on the side of trying to be verbose, but I can see how the false flag values add clutter. Now removed.

tyxla commented on client/reader/post-editor/index.jsx:
```diff
@@ -0,0 +1,78 @@
+import { serialize } from '@wordpress/blocks';
+import { Button } from '@wordpress/components';
+import { select, useDispatch } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
```
Is there a good reason to use `@wordpress/i18n` and not `i18n-calypso`?

creativecoder replied:
Nope, just habit. Now changed to use the Calypso package.

tyxla commented on client/data/posts/use-create-new-post.js:
```diff
@@ -0,0 +1,28 @@
+
+export default function useCreateNewPost( mutationOptions = {} ) {
+	const mutation = useMutation(
+		async ( { siteId, post } ) =>
```
The downside with this open `post` approach is that we don't really define a schema of what post data is expected here, and that is prone to errors and missing/unexpected data. Should we try to introduce some sort of schema, or at least document the expected data format better?

creativecoder replied:
I've moved the file to .ts and imported the relevant Post type from `@wordpress/core-data` -- let me know if that seems like a good solution.

tyxla commented on client/reader/post-editor/index.jsx:
```diff
@@ -0,0 +1,78 @@
+					<ToolbarSlot>
+						<Button
+							className="reader-post-editor__publish-button"
+							isPrimary="true"
```
Did you mean:

```suggestion
							isPrimary
```

creativecoder replied:
Yes, thanks!

tyxla commented on client/data/posts/use-create-new-post.js:
```diff
@@ -0,0 +1,28 @@
+import { useCallback } from 'react';
+import { useMutation } from 'react-query';
```
Nice to see React Query used for post creation!

Note that we might need to change the import statement since we're in the middle of upgrading the package to v4 in #74381.

creativecoder replied:
Thanks for flagging the react-query update!

tyxla commented on client/package.json:
```diff
@@ -40,6 +40,7 @@
 		"@automattic/happychat-connection": "workspace:^",
 		"@automattic/help-center": "workspace:^",
 		"@automattic/i18n-utils": "workspace:^",
+		"@automattic/isolated-block-editor": "2.22.0",
```
Is there a need to pin it like that, or could we make it less specific with a `^`?

creativecoder replied:
`^` doesn't work, since 0.23.0 updates to React 18 and so is incompatible. But I changed it to 0.22.x.

tyxla commented on client/reader/post-editor/index.jsx:
```diff
@@ -0,0 +1,78 @@
+
+		if ( blocks && blocks.length > 0 ) {
+			createNewPost( primarySiteId, {
+				content: serialize( blocks ),
```
Do we also need a post title?

creativecoder replied:
It's not technically required, though we'll want to handle this in some way, eventually.

tyxla commented on client/reader/post-editor/index.jsx:
```diff
@@ -0,0 +1,78 @@
+							className="reader-post-editor__publish-button"
+							isPrimary="true"
+							onClick={ publishPost }
+							disabled={ isLoading }
```
Should it also be disabled when there are no blocks?

creativecoder replied:
I had the same idea, but found this surprisingly tricky because `select( 'core/block-editor' ).getBlocks()` seems to only work if run within the scope of IsolatedBlockEditor, rather than within the parent component.

I didn't see an immediate way to pass a function that the button component runs to determine if it's disabled, rather than passing it a predetermined value.

I'll plan to address this in a follow-up, though!

tyxla commented on client/reader/following/main.jsx:
```diff
@@ -7,6 +9,7 @@ function FollowingStream( { ...props } ) {
 	/* eslint-disable wpcalypso/jsx-classname-namespace */
 	return (
 		<Stream { ...props }>
+			{ config.isEnabled( 'reader/editor' ) && <ReaderPostEditor /> }
```
Is this something we should async load? 

Otherwise, we'll bundle this for everyone, including production, even though we don't really render it.

creativecoder replied:
Great point, now updated to async load.

creativecoder commented:
I've rebased this branch off of `try-react-18` and updated `isolated-block-editor` to the latest version. It's not working yet, but there's less version forcing needed than before: pe7F0s-Vt-p2

bogiii commented:
I've reviewed the part of the code that owns the @Automattic/caribou team; it works as expected without regression.

Commit:
Reader: adds editor at top of feed behind a feature flag

creativecoder commented:
I've rebased this branch off of trunk. I wouldn't say the editor works, as there are some UI/style issues, but it builds and you can interact with it without errors.

worldomonation reviewed (changes requested):
I am not 100% sure why @Automattic/kitkat was tagged for review, but my concern is that this PR completely breaks Unit Tests.

noahtallen commented:
> my concern is that this PR completely breaks Unit Tests.

I think this will get resolved after this can be rebased with #78711
